File: src/types/video.ts
/**
 * Shared types for the LearnX AI Video Intelligence feature.
 *
 * These model the output of the AI processing pipeline described in the
 * product spec. In this frontend-only build the data is produced by
 * `src/data/videoIntelligenceMock.ts` instead of a real inference backend —
 * see `backend/` for the reference architecture these shapes map onto.
 */

export type PipelineStageId =
  | 'upload'
  | 'virus-scan'
  | 'metadata'
  | 'audio-extraction'
  | 'speech-detection'
  | 'vad'
  | 'diarization'
  | 'silence-detection'
  | 'scene-detection'
  | 'ocr'
  | 'transcription'
  | 'topic-detection'
  | 'chapter-detection'
  | 'concept-extraction'
  | 'summary'
  | 'flashcards'
  | 'quiz'
  | 'mindmap'
  | 'notes'
  | 'workspace-ready'

export type PipelineStageStatus = 'pending' | 'active' | 'done' | 'skipped' | 'error'

export interface PipelineStage {
  id: PipelineStageId
  label: string
  description: string
  status: PipelineStageStatus
  /** 0-100, only meaningful while status === 'active' */
  progress?: number
  durationMs?: number
}

export type VideoSourceType =
  | 'upload'
  | 'zoom'
  | 'teams'
  | 'meet'
  | 'screen-recording'
  | 'lecture'

export type VideoProcessingState = 'queued' | 'processing' | 'ready' | 'failed'

export interface SilenceSegment {
  id: string
  startSec: number
  endSec: number
  /** Why the AI flagged this window. */
  reason:
    | 'dead-air'
    | 'setup-time'
    | 'waiting'
    | 'repeated-pause'
    | 'idle-moment'
    | 'meaningful-pause'
  /** Whether this segment was cut from the optimized video. Meaningful pauses are always false. */
  removed: boolean
  confidence: number
}

export interface ChapterConcept {
  term: string
  definition: string
}

export interface Chapter {
  id: string
  index: number
  title: string
  startSec: number
  endSec: number
  difficulty: 'easy' | 'medium' | 'hard'
  confidence: number
  examImportance: number
  estimatedStudyMinutes: number
  keyConcepts: ChapterConcept[]
  formulas: string[]
  examTips: string[]
}

export interface TranscriptSegment {
  id: string
  startSec: number
  endSec: number
  speaker: string
  text: string
  chapterId: string
  highlighted?: boolean
}

export type SummaryLevel = 'quick' | 'detailed' | 'bullet' | 'exam' | 'revision' | 'one-minute'

export interface SummaryContent {
  level: SummaryLevel
  label: string
  points: string[]
}

export interface Flashcard {
  id: string
  chapterId: string
  question: string
  answer: string
  difficulty: 'easy' | 'medium' | 'hard'
  favorite: boolean
  masteredLevel: number // 0-5 spaced repetition box
}

export type QuizQuestionType = 'mcq' | 'true-false' | 'short-answer' | 'fill-blank'

export interface QuizQuestion {
  id: string
  chapterId: string
  type: QuizQuestionType
  prompt: string
  options?: string[]
  correctAnswer: string
  explanation: string
  difficulty: 'easy' | 'medium' | 'hard'
}

export interface MindMapNode {
  id: string
  label: string
  children: MindMapNode[]
}

export interface ChatCitation {
  chapterId: string
  chapterTitle: string
  timestampSec: number
}

export interface ChatMessage {
  id: string
  role: 'user' | 'assistant'
  text: string
  citations?: ChatCitation[]
}

export interface VideoStats {
  originalDurationSec: number
  optimizedDurationSec: number
  minutesSaved: number
  percentRemoved: number
  learningEfficiencyScore: number
}

export interface VideoLecture {
  id: string
  title: string
  course: string
  sourceType: VideoSourceType
  uploadedAt: string
  thumbnailGradient: [string, string]
  state: VideoProcessingState
  currentStageIndex: number
  pipeline: PipelineStage[]
  durationSec: number
  stats: VideoStats
  silenceSegments: SilenceSegment[]
  chapters: Chapter[]
  transcript: TranscriptSegment[]
  summaries: SummaryContent[]
  flashcards: Flashcard[]
  quiz: QuizQuestion[]
  mindMap: MindMapNode
  chat: ChatMessage[]
  /** Demo-only playable source; see component comments. */
  demoVideoUrl?: string
}

File: backend/app/schemas/video.py
"""
Pydantic response/request schemas for the AI Video Intelligence API.

These are a direct, field-for-field mirror of `src/types/video.ts` — kept
in sync intentionally so the frontend's mock-data shape and the real API
response shape never diverge, minimizing the future migration diff.
"""

from __future__ import annotations

from enum import Enum
from typing import Literal

from pydantic import BaseModel


class PipelineStageStatus(str, Enum):
    pending = "pending"
    active = "active"
    done = "done"
    skipped = "skipped"
    error = "error"


class PipelineStageOut(BaseModel):
    id: str
    label: str
    description: str
    status: PipelineStageStatus
    progress: float | None = None
    duration_ms: int | None = None


class SilenceSegmentOut(BaseModel):
    id: str
    start_sec: float
    end_sec: float
    reason: Literal[
        "dead-air", "setup-time", "waiting", "repeated-pause", "idle-moment", "meaningful-pause"
    ]
    removed: bool
    confidence: float


class ChapterConceptOut(BaseModel):
    term: str
    definition: str


class ChapterOut(BaseModel):
    id: str
    index: int
    title: str
    start_sec: float
    end_sec: float
    difficulty: Literal["easy", "medium", "hard"]
    confidence: float
    exam_importance: int
    estimated_study_minutes: int
    key_concepts: list[ChapterConceptOut]
    formulas: list[str]
    exam_tips: list[str]


class TranscriptSegmentOut(BaseModel):
    id: str
    start_sec: float
    end_sec: float
    speaker: str
    text: str
    chapter_id: str


class SummaryContentOut(BaseModel):
    level: Literal["quick", "detailed", "bullet", "exam", "revision", "one-minute"]
    label: str
    points: list[str]


class FlashcardOut(BaseModel):
    id: str
    chapter_id: str
    question: str
    answer: str
    difficulty: Literal["easy", "medium", "hard"]
    favorite: bool
    mastered_level: int


class QuizQuestionOut(BaseModel):
    id: str
    chapter_id: str
    type: Literal["mcq", "true-false", "short-answer", "fill-blank"]
    prompt: str
    options: list[str] | None
    correct_answer: str
    explanation: str
    difficulty: Literal["easy", "medium", "hard"]


class MindMapNodeOut(BaseModel):
    id: str
    label: str
    children: list["MindMapNodeOut"]


MindMapNodeOut.model_rebuild()


class VideoStatsOut(BaseModel):
    original_duration_sec: float
    optimized_duration_sec: float
    minutes_saved: float
    percent_removed: int
    learning_efficiency_score: int


class VideoLectureOut(BaseModel):
    id: str
    title: str
    course: str
    source_type: Literal["upload", "zoom", "teams", "meet", "screen-recording", "lecture"]
    uploaded_at: str
    state: Literal["queued", "processing", "ready", "failed"]
    current_stage_index: int
    pipeline: list[PipelineStageOut]
    duration_sec: float
    stats: VideoStatsOut
    silence_segments: list[SilenceSegmentOut]
    chapters: list[ChapterOut]


class VideoUploadInitRequest(BaseModel):
    """
    Starts a resumable/chunked upload session. The client then PUTs chunks
    to `/videos/{upload_id}/chunks/{index}` and finally POSTs
    `/videos/{upload_id}/complete`, at which point the pipeline is queued.
    """

    filename: str
    content_type: str
    total_size_bytes: int
    source_type: Literal["upload", "zoom", "teams", "meet", "screen-recording", "lecture"]
    course_id: str | None = None


class VideoUploadInitResponse(BaseModel):
    upload_id: str
    chunk_size_bytes: int
    chunk_count: int


class VideoChatRequest(BaseModel):
    message: str


class ChatCitationOut(BaseModel):
    chapter_id: str
    chapter_title: str
    timestamp_sec: float


class VideoChatResponse(BaseModel):
    """
    The assistant must ground every answer in retrieved transcript chunks
    (see app/services/rag.py) and always return at least one citation
    when the answer references lecture content — matching the product
    requirement "Never hallucinate. Always cite timestamps."
    """

    text: str
    citations: list[ChatCitationOut]
